### PR TITLE
Add animated transition beetween levels

### PR DESCRIPTION
**Pull request by JPCard**
**Please check if the PR fulfills these requirements:**

- [ x] The commit message follows our guidelines.
- For bug fixes and features:
    - [ x] You tested the changes.
    - [ ] You updated the docs or changelog.

Related issue (if applicable): #183 

**What kind of change does this PR introduce?**
This PR introduces an UI change.

**Does this PR introduce a breaking change?**
no, it doesn't.

## New feature or change ##

**What is the current behavior?** 
No animated transition beetween levels.

**What is the new behavior?**
Animated transition beetween levels.

**Other information**
There is a bug in fullscreen mode, the transition node doesn't cover up the entire screen.

### CODE REVIEW

File: game/src/Autoload/LevelLoader.gd [hunk 1]
```diff
@@ -36,11 +36,14 @@ func trigger(NewLevel: PackedScene, portal_name: String = "") -> void:
...
-#		yield(_game.transition, "peaked")
+	
 	
 	_game.level = _level
 	_game.add_child(_level)
 	_game.add_child(_player)
 	
+	if _game.transition.is_animating():
+		# This yield is surpassed 1 second before the transition animation is completed so the game can load the next level visually in the background
+		yield(_game.transition, "almost_completed")
```
[the remark this answers is not in the document]
Comment (JPCard) [reply game/src/Autoload/LevelLoader.gd]: I removed the arbitrary wait times and that signal.

File: game/src/Main/Game.tscn [hunk 2]
```diff
@@ -25,32 +29,158 @@ tracks/0/keys = {
 
 [sub_resource type="Animation" id=2]
 resource_name = "transition"
-length = 0.6
+length = 3.6
```
[the remark this answers is not in the document]
Comment (JPCard) [reply game/src/Main/Game.tscn]: I did that.

File: game/src/UI/Transition.gd [hunk 3]
```diff
@@ -1,6 +1,49 @@
 extends ColorRect
 
 
-signal peaked
+signal almost_completed
+
+const TRANSITION_ANIM_TOTAL_LENGTH := 3.6
```
[the remark this answers is not in the document]
Comment (JPCard) [reply game/src/UI/Transition.gd]: I don't use the animation duration anymore.

File: game/src/UI/Transition.gd [hunk 4]
```diff
@@ -1,6 +1,49 @@
...
+const TRANSITION_ANIM_MIN_LENGTH := 1.0
```
[the remark this answers is not in the document]
Comment (JPCard) [reply game/src/UI/Transition.gd]: I changed the animation so it's fade in and fade out. It only plays the loading animation if the loadout is taking too long.

File: game/src/UI/Transition.gd [hunk 5]
```diff
@@ -1,6 +1,49 @@
...
+const TRANSITION_ANIM_AMP := TRANSITION_ANIM_TOTAL_LENGTH - TRANSITION_ANIM_MIN_LENGTH
 
 onready var animation_player: AnimationPlayer = $AnimationPlayer
+
+
+
+func _ready() -> void:
+	randomize()
```
[the remark this answers is not in the document]
Comment (JPCard) [reply game/src/UI/Transition.gd]: I removed that.

File: game/src/UI/Transition.gd [hunk 6]
```diff
@@ -1,6 +1,49 @@
...
+
+
+func start_transition_animation() -> void:
+	# The animations will be transition_init->transition->transition_finalize
+	_change_animation("transition_init") 
+
+
+func _change_animation(anim_name: String) -> void:
```
[the remark this answers is not in the document]
Comment (JPCard) [reply game/src/UI/Transition.gd]: I used animation_player.play().

File: game/src/UI/Transition.gd [hunk 7]
```diff
@@ -1,6 +1,49 @@
...
+	var aux_length: float
+	aux_length = TRANSITION_ANIM_MIN_LENGTH + rand_range(0,TRANSITION_ANIM_AMP)
+	# This while makes imposible to end the animation when flipping_h the player art sprite
+	while( (aux_length >= TRANSITION_ANIM_TOTAL_LENGTH * 0.5) and (aux_length < TRANSITION_ANIM_TOTAL_LENGTH / 0.5 + 0.2) ):
+		aux_length = TRANSITION_ANIM_MIN_LENGTH + rand_range(0,TRANSITION_ANIM_AMP)
+	animation_player.get_animation("transition").length = aux_length
+
+
+func _on_AnimationPlayer_animation_finished(anim_name: String) -> void:
+	if(anim_name == "transition_init"):
```
[the remark this answers is not in the document]
Comment (JPCard) [reply game/src/UI/Transition.gd]: I used the command line tool, thanks for the advice.

File: game/src/UI/Transition.gd [hunk 8]
```diff
@@ -1,6 +1,49 @@
...
+	# The animations will be transition_init->transition->transition_finalize
+	_change_animation("transition_init") 
+
+
+func _change_animation(anim_name: String) -> void:
+	animation_player.play(anim_name)
+
+
+func _change_transition_length() -> void:
+	var aux_length: float
```
[the remark this answers is not in the document]
Comment (JPCard) [reply game/src/UI/Transition.gd]: I removed _change_transition_length().

File: game/src/UI/Transition.gd [hunk 9]
```diff
@@ -1,6 +1,49 @@
...
+	elif(anim_name == "transition"):
+		_change_animation("transition_finalize")
+		# This yield gives 1 second to go to the next level in the background to keep the transition_finalize end clean.
+		yield(get_tree().create_timer( animation_player.get_animation("transition_finalize").length - 1), "timeout") 
+		emit_signal("almost_completed")
+
+
+func is_animating() -> bool: 
+	# <BASE> is the "do nothing" animation
+	return animation_player.current_animation != "<BASE>"
```
[the remark this answers is not in the document]
Comment (JPCard) [reply game/src/UI/Transition.gd]: I removed is_animating().

**Comment by NathanLovato** [comment]
Thanks for the changes. There are some more issues, like the fade not working full screen, and some code cleanup, but I'll take care of it and merge manually from there.

**Comment by NathanLovato** [comment]
Merged manually in 91e0659b8c1ed122b749facbe197b6ebbe0fface

**Comment by NathanLovato** [comment]
Thanks for your contribution! Your code needed some more simplifications. You can see everything I changed here: 2742ee66e8948fa92585e27fa5f76a55a1748d59

For future contributions, I'd recommend you to create a git branch to have your work and to not merge the way you did. Add new changes as new commits in the same branch or amend your initial commit (git commit --amend) and force push (git push -f). You shouldn't have 14 commits with multiple merges. It will also make my job easier.

Also, you don't need to reply to every review comment. Just request a new review once your code is ready to review or send me a message and I'll get to it!

Thanks again, I'll close this PR but your changes are in the repository.

**Pull request closed without merging**